pygfx/renderers/wgpu/_pipeline.py

```diff
@@ -0,0 +1,758 @@
+                    "stencil_front": {},  # use defaults
+                    "stencil_back": {},  # use defaults
+                },
+                multisample={
+                    "count": 1,
+                    "mask": 0xFFFFFFFF,
+                    "alpha_to_coverage_enabled": False,
+                },
+                fragment={
+                    "module": shader_module,
+                    "entry_point": "fs_main",
+                    "targets": color_descriptors,
+                },
+            )
+
+        return pipelines
+
+    def draw(self, render_pass, environment, pass_index, render_mask):
+        """Draw the pipeline, doing the actual rendering job."""
+        if self.broken:
+            return
+
+        if not (render_mask & self.render_info["render_mask"]):
+            return
+        env_hash = environment.hash
+
+        # Collect what's needed
+        pipeline = self.wgpu_pipelines[env_hash][pass_index]
+        indices = self.render_info["indices"]
+        index_buffer = self.resources["index_buffer"]
+        vertex_buffers = self.resources["vertex_buffers"]
+        bind_groups = self.wgpu_bind_groups
+
+        # Set pipeline and resources
+        render_pass.set_pipeline(pipeline)
+        for slot, vbuffer in vertex_buffers.items():
+            render_pass.set_vertex_buffer(
+                slot,
+                vbuffer._wgpu_buffer[1],
+                vbuffer.vertex_byte_range[0],
+                vbuffer.vertex_byte_range[1],
+            )
+        for bind_group_id, bind_group in enumerate(bind_groups):
+            render_pass.set_bind_group(bind_group_id, bind_group, [], 0, 99)
+
+        # Draw!
+        # draw_indexed(count_v, count_i, first_vertex, base_vertex, first_instance)
+        # draw(count_vertex, count_instance, first_vertex, first_instance)
+        if index_buffer is not None:
+            render_pass.set_index_buffer(index_buffer, 0, index_buffer.size)
+            if len(indices) == 4:
+                base_vertex = 0  # A value added to each index before reading [...]
+                indices = list(indices)
+                indices.insert(-1, base_vertex)
+            render_pass.draw_indexed(*indices)
+        else:
+            render_pass.draw(*indices)
+
+
+class Cache:
+    def __init__(self):
+        self._d = {}
+
+    def get(self, key):
+        return self._d.get(key, None)
+
+    def set(self, key, value):
+        self._d[key] = value
+
+    def get_shader_module(self, device, source):
+        """Compile a shader module object, or re-use it from the cache."""
+        # todo: also release shader modules that are no longer used
+        # todo: cache more objects, like pipelines once we figure out how to clean things up
+
+        assert isinstance(source, str)
+        key = source  # or hash(code)
+
+        m = self.get(key)
+        if m is None:
+            m = device.create_shader_module(code=source)
+            self.set(key, m)
```

Comment:
is it possible that a different device is used for the same source, and that that leads to problems?

Reply:
No, because in pygfx we only use a single device.